### PR TITLE
pkg/config: use generic pipeline function

### PR DESCRIPTION
**Pull request by bbguimaraes**
This pattern is everywhere, implemented slightly differently each time.  Some
versions do not bound the maximum number of Go routines, causing file descriptor
exhaustion.

This change creates a small library that is as generic as I managed to make it
in Go and reimplements `ci-operator` configuration loading using it.  More to
come.

/hold

Based on https://github.com/openshift/ci-tools/pull/2627.

https://issues.redhat.com/browse/DPTP-2531

---

See also:

- [ ] https://github.com/openshift/ci-tools/pull/2629
- [ ] https://github.com/openshift/ci-tools/pull/2630
- [ ] https://github.com/openshift/ci-tools/pull/2631
- [ ] https://github.com/openshift/ci-tools/pull/2632

### CODE REVIEW

File: pkg/util/sync_test.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package util
+
+import (
+	"fmt"
+)
+
+func ExampleProduceMapReduce() {
```
Comment (petr-muller) [inline pkg/util/sync_test.go:7]: Can we make this a real test?

Comment (bbguimaraes) [reply pkg/util/sync_test.go:7]: It is a real test:

```
$ go test -v ./pkg/util | grep Example
=== RUN   ExampleProduceMapReduce
--- PASS: ExampleProduceMapReduce (0.00s)
```

and is [supposed to be how examples are embedded in package documentation](https://go.dev/blog/examples), which is why I decided to try to write it this way.

Comment (bbguimaraes) [reply pkg/util/sync_test.go:7]: After the usual fight (I hate Go modules), this is how it looks like rendered:

<details>

Screen Shot 2022-02-09 at 12 50 29

</details>

Apparently, there is no way to format `code`, ` `` ` is not respected and I can't find examples in the stdlib.

Comment (petr-muller) [reply pkg/util/sync_test.go:7]: TIL

Comment (petr-muller) [reply pkg/util/sync_test.go:7]: Pretty awesome

**Review by petr-muller — APPROVED** [review]

**Pull request merged**